scio-extra/src/main/scala/com/spotify/scio/extra/sparkey/ShardedSparkeyUri.scala

```diff
@@ -0,0 +1,157 @@
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.spotify.scio.extra.sparkey
+
+import java.io.File
+import java.net.URI
+
+import com.spotify.scio.util.{RemoteFileUtil, ScioUtil}
+import com.spotify.sparkey.SparkeyReader
+import com.spotify.sparkey.extra.ThreadLocalSparkeyReader
+import org.apache.beam.sdk.io.FileSystems
+import org.apache.beam.sdk.io.fs.MatchResult
+import org.apache.beam.sdk.options.PipelineOptions
+
+import scala.collection.JavaConverters._
+
+/**
+ * Sharded Sparkey support
+ */
+trait ShardedSparkeyUri extends SparkeyUri {
+  val basePath: String
+  def getReader: ShardedSparkeyReader
+
+  private[sparkey] def exists: Boolean
+  override def toString: String = basePath
+
+  def basePathForShard(shardIndex: Short, numShards: Short): String =
+    f"$basePath/part-$shardIndex%05d-of-$numShards%05d"
+
+  def sparkeyUriForShard(shardIndex: Short, numShards: Short): SparkeyUri
+
+  lazy val globExpression = s"$basePath/part-*"
+
+  private[sparkey] def basePathsAndCount(): (Seq[String], Short) = {
+    val matchResult: MatchResult = FileSystems.`match`(globExpression)
+    val paths = matchResult.metadata().asScala.map(_.resourceId.toString)
+    val indexPaths = paths.filter(_.endsWith(".spi")).sorted
+
+    val allStartParts = indexPaths.map(ShardedSparkeyUri.shardIndexFromPath)
+    val allEndParts = indexPaths.map(ShardedSparkeyUri.numShardsFromPath)
+
+    val distinctNumShards = allEndParts.toSet
+    if (distinctNumShards.isEmpty) {
+      (Seq.empty[String], 0)
+    } else {
+      require(
+        distinctNumShards.size == 1,
+        s"Expected all .spi files to end with the same shard count, but found: $distinctNumShards."
+      )
+
+      val numShards = distinctNumShards.iterator.next
+
+      val numShardFiles = allStartParts.toSet.size
+      require(
+        numShardFiles <= numShards,
+        "Expected the number of Sparkey shards to be less than or equal to the " +
+          s"total shard count ($numShards), but found $numShardFiles"
+      )
+
+      val basePaths = indexPaths.map(_.replaceAll("\\.spi$", ""))
+
+      (basePaths, numShards)
+    }
+  }
+}
+
+private[sparkey] object ShardedSparkeyUri {
+  def apply(basePath: String, options: PipelineOptions): ShardedSparkeyUri =
+    if (ScioUtil.isLocalUri(new URI(basePath))) {
+      LocalShardedSparkeyUri(basePath)
+    } else {
+      RemoteShardedSparkeyUri(basePath, RemoteFileUtil.create(options))
+    }
+
+  private[sparkey] def numShardsFromPath(path: String): Short =
+    path.split("-of-").toList.last.split("\\.").head.toShort
+
+  private[sparkey] def shardIndexFromPath(path: String): Short =
+    path.split("part-").toList.last.split("-of-").head.toShort
+
+  private[sparkey] def localReadersByShard(
+    localBasePaths: Iterable[String]
+  ): Map[Short, SparkeyReader] = {
+    localBasePaths
+      .map(
```

Comment:
🎨 I think that in this case `{}` is better suited

Reply:
also not that it's huge deal here but `localBasePaths.iterator.map` to avoid extra copying